src/native/credentials_provider.c

```diff
@@ -268,6 +273,96 @@ JNIEXPORT jlong JNICALL
+    jbyteArray java_secret_access_key =
+        (*env)->GetObjectField(env, java_credentials, credentials_properties.secret_access_key_field_id);
+    jbyteArray java_session_token =
+        (*env)->GetObjectField(env, java_credentials, credentials_properties.session_token_field_id);
+
+    struct aws_byte_cursor access_key_id = aws_jni_byte_cursor_from_jbyteArray_acquire(env, java_access_key_id);
+    struct aws_byte_cursor secret_access_key = aws_jni_byte_cursor_from_jbyteArray_acquire(env, java_secret_access_key);
+    struct aws_byte_cursor session_token = aws_jni_byte_cursor_from_jbyteArray_acquire(env, java_session_token);
+    struct aws_credentials *native_credentials =
+        aws_credentials_new(allocator, access_key_id, secret_access_key, session_token, UINT64_MAX);
+    if (!native_credentials) {
+        aws_jni_throw_runtime_exception(env, "Failed to create native credentials");
+        // error has been raised from creating function
+        goto done;
+    }
+    callback(native_credentials, AWS_ERROR_SUCCESS, callback_user_data);
+    aws_credentials_release(native_credentials);
+
+    return_value = AWS_OP_SUCCESS;
+done:
+    aws_jni_byte_cursor_from_jbyteArray_release(env, java_access_key_id, access_key_id);
+    aws_jni_byte_cursor_from_jbyteArray_release(env, java_secret_access_key, secret_access_key);
+    aws_jni_byte_cursor_from_jbyteArray_release(env, java_session_token, session_token);
+call_failed:
+    (*env)->DeleteLocalRef(env, java_credentials);
+    return return_value;
+}
+
+JNIEXPORT jlong JNICALL
+    Java_software_amazon_awssdk_crt_auth_credentials_DelegateCredentialsProvider_delegateCredentialsProviderNew(
+        JNIEnv *env,
+        jclass jni_class,
+        jobject java_crt_credentials_provider,
+        jobject jni_delegate_credential_handler) {
+
+    (void)jni_class;
+    (void)env;
+
+    struct aws_allocator *allocator = aws_jni_get_allocator();
+    struct aws_credentials_provider_callback_data *callback_data =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_credentials_provider_callback_data));
+    callback_data->java_crt_credentials_provider = (*env)->NewWeakGlobalRef(env, java_crt_credentials_provider);
+    callback_data->jni_delegate_credential_handler = (*env)->NewGlobalRef(env, jni_delegate_credential_handler);
+
+    struct aws_credentials_provider_delegate_options options = {
+        .get_credentials = s_credentials_provider_delegate_get_credentials,
+        .delegate_user_data = callback_data,
+        .shutdown_options =
+            {
+                .shutdown_callback = s_on_shutdown_complete,
+                .shutdown_user_data = callback_data,
+            },
+    };
+
+    jint jvmresult = (*env)->GetJavaVM(env, &callback_data->jvm);
+    AWS_FATAL_ASSERT(jvmresult == 0);
+
+    struct aws_credentials_provider *provider = aws_credentials_provider_new_delegate(allocator, &options);
+    if (provider == NULL) {
+        aws_mem_release(allocator, callback_data);
```

Comment:
See above comment, we need to fully clean up callback_data since it holds multiple refs potentially at this point.